src/routes/getVideoLabel.ts

```diff
@@ -0,0 +1,176 @@
+                };
+
+                acc[segment.videoID].segments ??= [];
+                acc[segment.videoID].segments.push(segment);
+
+                return acc;
+            }, {});
+
+        for (const [videoID, videoData] of Object.entries(segmentPerVideoID)) {
+            const data: VideoData = {
+                hash: videoData.hash,
+                segments: chooseSegment(videoData.segments),
+            };
+
+            if (data.segments.length > 0) {
+                segments[videoID] = data;
+            }
+        }
+
+        return segments;
+    } catch (err) {
+        Logger.error(err as string);
+        return null;
+    }
+}
+
+async function getSegmentsFromDBByHash(hashedVideoIDPrefix: VideoIDHash, service: Service): Promise<DBSegment[]> {
+    const fetchFromDB = () => db
+        .prepare(
+            "all",
+            `SELECT "startTime", "endTime", "videoID", "votes", "locked", "UUID", "userID", "category", "actionType", "hashedVideoID", "description" FROM "sponsorTimes"
+            WHERE "hashedVideoID" LIKE ? AND "service" = ? AND actionType = 'full' AND "hidden" = 0 AND "shadowHidden" = 0`,
+            [`${hashedVideoIDPrefix}%`, service]
+        ) as Promise<DBSegment[]>;
+
+    if (hashedVideoIDPrefix.length === 4) {
+        return await QueryCacher.get(fetchFromDB, videoLabelsHashKey(hashedVideoIDPrefix, service));
+    }
+
+    return await fetchFromDB();
+}
+
+async function getSegmentsFromDBByVideoID(videoID: VideoID, service: Service): Promise<DBSegment[]> {
+    const fetchFromDB = () => db
+        .prepare(
+            "all",
+            `SELECT "startTime", "endTime", "votes", "locked", "UUID", "userID", "category", "actionType", "description" FROM "sponsorTimes" 
+            WHERE "videoID" = ? AND "service" = ? AND "hidden" = 0 AND "shadowHidden" = 0`,
+            [videoID, service]
+        ) as Promise<DBSegment[]>;
+
+    return await QueryCacher.get(fetchFromDB, videoLabelsKey(videoID, service));
+}
+
+function chooseSegment<T extends DBSegment>(choices: T[]): Segment[] {
+    // filter out -2 segments
+    choices = choices.filter((segment) => segment.votes > -2);
+    const results = [];
+    // trivial decisions
+    if (choices.length === 0) {
+        return [];
+    } else if (choices.length === 1) {
+        return transformDBSegments(choices);
+    }
+    // if locked, only choose from locked
+    const locked = choices.filter((segment) => segment.locked);
+    if (locked.length > 0) {
+        choices = locked;
+    }
+    //no need to filter, just one label
+    if (choices.length === 1) {
+        return transformDBSegments(choices);
+    }
+    // sponsor > exclusive > selfpromo
+    const sponsorResult = choices.find((segment) => segment.category === "sponsor");
+    const eaResult = choices.find((segment) => segment.category === "exclusive_access");
+    const selfpromoResult = choices.find((segment) => segment.category === "selfpromo");
```

Comment:
maybe convert these to lambda functions to only run as many as are needed

Reply:
```ts
 const categoryResult = findCategory("sponsor") ?? findCategory("exclusive_access") ?? findCategory("selfpromo");
    if (categoryResult) results.push(categoryResult);
```